core/src/ipc.rs

```diff
@@ -131,3 +135,99 @@ impl<T> FromRawFd for IpcReceiver<T> {
         }
     }
 }
+
+/// Creates a pair of sockets that are ment for passing file descriptors to IO resources of type T.
+pub fn io_pair<T>() -> TypedResult<(IoSender<T>, IoReceiver<T>)>
+where
+    T: AsRawFd,
+{
+    let (tx, rx) = socketpair(
+        AddressFamily::Unix,
+        SockType::Datagram,
+        None,
+        SockFlag::SOCK_NONBLOCK,
+    )
+    .typ(SystemError::Panic)?;
+    unsafe { Ok((IoSender::from_raw_fd(tx), IoReceiver::from_raw_fd(rx))) }
+}
+
+#[derive(Debug)]
+/// Internal data type for the IO resource sender
+pub struct IoSender<T> {
+    socket: UnixDatagram,
+    _p: PhantomData<T>,
+}
+
+impl<T> IoSender<T>
+where
+    T: AsRawFd,
+{
+    pub fn try_send(&self, fds: &[T]) -> TypedResult<()> {
+        let fds: Vec<RawFd> = fds.iter().map(T::as_raw_fd).collect();
+        let cmsg = [ControlMessage::ScmRights(&fds)];
+        let iov = [IoSlice::new(&[0u8])];
+        let fd = self.socket.as_raw_fd();
+        sendmsg::<()>(fd, &iov, &cmsg, MsgFlags::empty(), None).typ(SystemError::Panic)?;
+        Ok(())
+    }
+}
+
+impl<T> IoReceiver<T>
+where
+    T: FromRawFd,
+{
+    pub fn try_receive(&self) -> TypedResult<T> {
+        let mut cmsg = cmsg_space!(RawFd);
+        let mut iobuf = [0u8];
+        let mut iov = [IoSliceMut::new(&mut iobuf)];
+        let fd = self.socket.as_raw_fd();
+        let msg = recvmsg::<()>(fd, &mut iov, Some(&mut cmsg), MsgFlags::empty())
+            .typ(SystemError::Panic)?;
+        for cmsg in msg.cmsgs() {
+            if let ControlMessageOwned::ScmRights(fds) = cmsg {
+                let t = fds.iter().map(|fd| unsafe { T::from_raw_fd(*fd) }).next();
+                if let Some(t) = t {
+                    return Ok(t);
+                }
+            }
+        }
+        Err(SystemError::Panic).typ(SystemError::Panic)
+    }
+}
+
+#[derive(Debug)]
+/// Internal data type for the IO resource sender
+pub struct IoReceiver<T> {
+    socket: UnixDatagram,
+    _p: PhantomData<T>,
+}
+
+impl<T> AsRawFd for IoSender<T> {
+    fn as_raw_fd(&self) -> RawFd {
+        self.socket.as_raw_fd()
+    }
+}
+
+impl<T> AsRawFd for IoReceiver<T> {
+    fn as_raw_fd(&self) -> RawFd {
+        self.socket.as_raw_fd()
+    }
+}
+
+impl<T> FromRawFd for IoSender<T> {
+    unsafe fn from_raw_fd(fd: RawFd) -> Self {
+        Self {
+            socket: unsafe { UnixDatagram::from_raw_fd(fd) },
```

Comment:
This looks like something that could fail, because fd may not be a valid unix socket fd. Maybe the `try_from` trait would make more sense here?

Reply:
How would we detect that it fails? I don't think this is possible, and this is why this is marked as `unsafe`. See also `IpcSender`.